track users across registration, matching anon and signed in events in mixpanel [BW-409]

Shakespeared opened the pull request.

ehanna4 committed:
WIP

ehanna4 committed:
Preliminaryly working anon event track on page views WIP on merge

Co-Authored-By: Kai <19541449+Shakespeared@users.noreply.github.com>

Shakespeared committed:
refactor Register to be able to use hooks

Co-Authored-By: ehanna4 <54322292+ehanna4@users.noreply.github.com>
Co-Authored-By: Scout Rubenstein <15351021+s-rubenstein@users.noreply.github.com>

Shakespeared committed:
update unregistereduseridstore to use get()

Co-Authored-By: Scout Rubenstein <15351021+s-rubenstein@users.noreply.github.com>

Shakespeared committed:
fix lint error

s-rubenstein committed:
Working on changing the anonymousId to the authStore

Shakespeared committed:
code cleanup and refactor anonymous id tracking

Co-Authored-By: ehanna4 <54322292+ehanna4@users.noreply.github.com>
Co-Authored-By: Brett Heath-Wlaz <555632+panentheos@users.noreply.github.com>
Co-Authored-By: Scout Rubenstein <15351021+s-rubenstein@users.noreply.github.com>

ehanna4 committed:
remove unused things

panentheos committed:
tweaks

ehanna4 commented:
In testing today I am noticing an identify event in mixpanel whenever I am already logged in then close Terra without signing out and re-open a Terra window.

ehanna4 committed:
Only registering users should call identify endpoint

due to 500 cap we cannot have identify being called on every page refresh/login/etc.

Co-Authored-By: Kai <19541449+Shakespeared@users.noreply.github.com>
Co-Authored-By: Scout Rubenstein <15351021+s-rubenstein@users.noreply.github.com>

panentheos commented on src/libs/auth.js:
```diff
@@ -197,6 +198,14 @@ authStore.subscribe(async (state, oldState) => {
 })
 
+authStore.subscribe(async (state, oldState) => {
+  if (oldState.registrationStatus === 'unregistered' && state.registrationStatus === 'registered') {
```
TOL: This does prevent extra `identify` calls by only triggering on registration, but there's still an underlying issue we might want to address at some point, specifically that on initial page load, if the path is public, the page view event will fire before the user's status is confirmed, so it will be 'orphaned' as an anonymous event.

A potential fix would be to modify the logic in `AuthContainer` to not render public pages as eagerly. The tradeoff in that case would be slightly longer perceived load time for registered users.

Shakespeared replied:
What do you think about going ahead with that change vs merging as it is now, or maybe leaving a comment or TODO there to indicate a possible future improvement?

panentheos replied:
I think a comment would be good, perhaps in `PageViewReporter` since that's what ultimately gets affected.

ehanna4 replied:
I just chatted with Amelia and the main question they are looking to get at is the conversion rate from people who browse anonymously and leave vs people who then register to use the site. So I think it might be worthwhile to try @panentheos 's suggestion in this PR so we don't muddy the data with orphaned anon events waiting to add it later

panentheos reviewed (dismissed):
Looks good, tested locally by running the integration test.

panentheos committed:
Merge remote-tracking branch 'origin/dev' into kf-eh-anon-events

panentheos committed:
revert export rename

panentheos committed:
rename mergeEvent to identify and call on logins as well

panentheos committed:
document order of operation

panentheos commented:
This is ready for a fresh look. The issue regarding page loads has been addressed by the addition of `ensureAuthSettled`, so we now avoid recording/identifying anonymous events unless the user is actually signed out. Tested locally.

salonishah11 reviewed (approved):
LGTM 👍

Shakespeared commented:
This update looks good to me, tested locally. Thanks for resolving the page view issue!
